utils/providers.py

```diff
@@ -1,388 +1,391 @@
+                    return False
+        except Exception:  # No SSH connection
+            return False
+        return True
+
+    def __call__(self, provider):
+        """ When an instance of this class is called, it will apply this filter on a given provider
+
+        Usage:
+            pf = ProviderFilter('cloud_infra', categories=['cloud', 'infra'])
+            providers = list_providers([pf])
+            pf2 = ProviderFilter(types=['openstack', 'ec2'], required_fields=['small_template'])
+            provider_keys = list_provider_keys([pf, pf2])
+            ...or...
+            pf = ProviderFilter(required_tags=['openstack', 'complete'])
+            pf_inverted = ProviderFilter(required_tags=['disabled'], inverted=True)
+            providers = setup_a_provider([pf, pf_inverted])
+
+        Returns:
+            `True` if provider passed all checks and was not filtered out, `False` otherwise.
+            The result is opposite if the 'inverted' attribute is set to `True`.
+        """
+        if self.keys and provider.key not in self.keys or \
+           self.categories is not None and provider.category not in self.categories or \
+           self.types is not None and provider.type not in self.types or \
+           self.required_fields and not self._filter_required_fields(provider,
+                                                                     self.required_fields) or \
+           self.restrict_version and not self._filter_restricted_version(provider) or \
+           self.required_tags and not self._filter_tags(provider, self.required_tags) or \
+           self.rejected_tags and self._filter_tags(provider, self.rejected_tags):
+            return self.inverted
+        return not self.inverted
+
+    def copy(self):
+        return copy(self)
+
+
+global_filters['enabled_only'] = ProviderFilter(required_tags=['disabled'], inverted=True)
+
+
+def list_providers(filters=None, use_global_filters=True):
+    """ Returns list of provider crud objects, optional filtering (enabled by default)
+
+    Args:
+        filters: List if :py:class:`ProviderFilter` or None
+        use_global_filters: Will apply global filters as well if `True`, will not otherwise
     """
-    if not allowed_types:
-        allowed_types = [
-            k2 for k in BaseProvider.type_mapping.values() for k2 in k.provider_types.keys()
-        ]
-    providers = []
-    for provider, data in providers_data.items():
-        provider_type = data.get("type", None)
-        if provider not in filtered:
-            continue
-        assert provider_type is not None, "Provider {} has no type specified!".format(provider)
-        if provider_type in allowed_types:
-            providers.append(provider)
+    filters = filters or []
+    if use_global_filters:
+        filters = filters + global_filters.items()
+    providers = [get_crud(prov_key) for prov_key in providers_data]
```

Comment:
Given what we were saying earlier about late importing the provider objects, this section here will import all provider objects for all providers that appear management_systems. So we would already have imported all modules. I feel we should take the opportunity and change the name of this variable also ```management_systems``` as opposed to ```providers_data``` so that then at a glance there is a direct correlation between the var and the yaml.